proxy/http2/Http2ConnectionState.cc

```diff
@@ -1427,31 +1477,65 @@ Http2ConnectionState::restart_streams()
 void
 Http2ConnectionState::restart_receiving(Http2Stream *stream)
 {
-  uint32_t initial_rwnd = this->local_settings.get(HTTP2_SETTINGS_INITIAL_WINDOW_SIZE);
-  uint32_t min_rwnd     = std::min(initial_rwnd, this->local_settings.get(HTTP2_SETTINGS_MAX_FRAME_SIZE));
-
   // Connection level WINDOW UPDATE
-  if (this->server_rwnd() < min_rwnd) {
-    Http2WindowSize diff_size = initial_rwnd - this->server_rwnd();
-    this->increment_server_rwnd(diff_size);
-    this->_server_rwnd_is_shrinking = false;
-    this->send_window_update_frame(0, diff_size);
+  uint32_t const configured_session_window = this->_get_configured_receive_session_window_size_in();
+  uint32_t const min_session_window = std::min(configured_session_window, this->local_settings.get(HTTP2_SETTINGS_MAX_FRAME_SIZE));
+  if (this->server_rwnd() < min_session_window) {
+    Http2WindowSize diff_size = configured_session_window - this->server_rwnd();
+    if (diff_size > 0) {
+      this->increment_server_rwnd(diff_size);
+      this->_server_rwnd_is_shrinking = false;
+      this->send_window_update_frame(HTTP2_CONNECTION_CONTROL_STREAM, diff_size);
+    }
   }
 
   // Stream level WINDOW UPDATE
-  if (stream == nullptr || stream->server_rwnd() >= min_rwnd) {
+  if (stream == nullptr || stream->server_rwnd() >= min_session_window) {
+    // There's no need to increase the stream window size if it is already big
+    // enough to hold what the stream/max frame size can receive.
     return;
   }
 
+  uint32_t initial_stream_window = this->local_settings.get(HTTP2_SETTINGS_INITIAL_WINDOW_SIZE);
+  if (this->_has_dynamic_stream_window()) {
+    // If our stream window size is dynamic, then we have to allow for this situation:
+    //
+    // 1. ATS initializes the stream window size to the configured value via a
+    // SETTINGS frame, say 10 bytes.
+    //
+    // 2. The peer sends a data frame of that size, again 10 bytes.
+    //
+    // 3. At the same time that the peer sends the data frame, we shrink the
+    // stream window size via a SETTINGS frame to a smaller value, say 5 bytes.
+    //
+    // 4. The peer receives the SETTINGS frame and calculates its send window
+    // to be -5 bytes.
+    //
+    // 5. ATS cannot simply send a WINDOW_UPDATE frame of 5 bytes to bring the
```

Comment:
ATS gave 10 at the beginning and the peer already consumed 10. If ATS has already processed the 10 bytes received, we should be able to give 5. If ATS has not processed the 10 bytes, there's nothing to do here yet. SETTINGS frame doesn't affect any existing streams on the peer side.

Reply:
We discussed this offline together, but just to record it here: the RFC states that existing stream window sizes (as well as new ones) do get adjusted per SETTINGS frames that configure SETTINGS_INITIAL_WINDOW_SIZE:

https://www.rfc-editor.org/rfc/rfc9113.html#name-initial-flow-control-window

> In addition to changing the flow-control window for streams that are not yet active, a [SETTINGS](https://www.rfc-editor.org/rfc/rfc9113.html#SETTINGS) frame can alter the initial flow-control window size for streams with active flow-control windows (that is, streams in the “open” or “half-closed (remote)” state). When the value of [SETTINGS_INITIAL_WINDOW_SIZE](https://www.rfc-editor.org/rfc/rfc9113.html#SETTINGS_INITIAL_WINDOW_SIZE) changes, a receiver MUST adjust the size of all stream flow-control windows that it maintains by the difference between the new value and the old value.[¶](https://www.rfc-editor.org/rfc/rfc9113.html#section-6.9.2-3)

None of this is a problem: the math actually works out pretty seamlessly for negative receive windows.